### PR TITLE
Fix Knockback

### PR DESCRIPTION
**Pull request by kphoenix137**
This fixes the problem with knockback. Knockback always pushes monsters in the opposite direction that the monster is facing. This can result in monsters being pushed in the wrong direction if they aren't facing the attacking player/missile.

Melee attacks and Telekinesis will calculate a direction based on the player tile and monster future tile upon initiating knockback.

Missiles will calculate a direction based on the missile start tile and monster future tile upon initiating knockback.

### CODE REVIEW

**Comment by AJenbo** [comment]
Are they also to fast for you to turn around?

Was it not possible to calculate the arrow direction?

**Comment by kphoenix137** [comment]
Maybe player facing direction was a bit inaccurate. It calculates direction using the player occupied tile and monster occupied tile? Which generally is whatever direction the player is facing.

**Comment by AJenbo** [comment]
> Would that matter? Isn't it based on your current position, not which way you are facing?

I only read the description so fare, I was mislead apparently.

**Comment by kphoenix137** [comment]
> > Would that matter? Isn't it based on your current position, not which way you are facing?
> 
> I only read the description so fare, I was mislead apparently.

I just revised it to solve the potential edge case in the first post. I will revise the first post as well.

File: Source/monster.cpp [hunk 1]
```diff
@@ -3733,9 +3733,9 @@ void M_ClearSquares(const Monster &monster)
 	}
 }
 
-void M_GetKnockback(Monster &monster)
+void M_GetKnockback(Monster &monster, WorldTilePosition attackerStartPos)
 {
-	Direction dir = Opposite(monster.direction);
+	auto dir = GetDirection(attackerStartPos, monster.position.tile);
```
Comment (AJenbo) [inline Source/monster.cpp:3738]:
```suggestion
	Direction dir = GetDirection(attackerStartPos, monster.position.tile);
```

File: Source/player.cpp [hunk 2]
```diff
@@ -677,8 +677,8 @@ bool PlrHitMonst(Player &player, Monster &monster, bool adjacentDamage = false)
 	if ((monster.hitPoints >> 6) <= 0) {
 		M_StartKill(monster, player);
 	} else {
-		if (monster.mode != MonsterMode::Petrified && HasAnyOf(player._pIFlags, ItemSpecialEffect::Knockback))
-			M_GetKnockback(monster);
+		if (monster.mode != MonsterMode::Petrified && HasAnyOf(player._pIFlags, ItemSpecialEffect::Knockback) && &player == MyPlayer)
+			NetSendCmdParam5(true, CMD_KNOCKBACK, monster.getId(), player.position.tile.x, player.position.tile.y, 0, 0);
```
Comment (AJenbo) [inline Source/player.cpp:681]: Won't this change delay the effect by 1 frame, and why is it needed?

**Review by AJenbo — APPROVED** [review]

**Pull request merged**